Workiva/over_react
# FED-1863 Required props lint

## Dependent PR

This PR is dependent on https://github.com/Workiva/over_react/pull/870

---

## Motivation
To help provide a good developer experience when working with required late props, we want to provide a lint that warns when a component is missing required props.

That way, devs can get static warnings, and not wait until they hit our runtime errors to find out they were missing props.

This lint should be similar to the warnings you get when you are missing required function parameters.

We had a partial implementation of this lint, and we needed to get it production-ready.

## Changes
- Finish `MissingRequiredPropDiagnostic` implementation, enable it, add tests
    - This supports both `late` required and `@requiredProp` required props, but after some back and forth I opted to only lint for `late` required by default. For a more detailed explanation, see comment in tools/analyzer_plugin/lib/src/plugin.dart
- Add utilities used by the diagnostic, with unit tests
  - `getAllProps` to get all props for a props class and its supertypes
  - `getAllRequiredProps` to collect info on which props are required
  - `isRequiredPropValidationDisabled`
- Raise analyzer plugin SDK language constraint to 2.17 to use enhanced enums
- Other changes not directly related to this diagnostic
    - Work around InconsistentAnalysisExceptions when typing
        - I noticed when manually testing that we'd get a bunch errors in the IDE about `InconsistentAnalysisException`, unrelated to this PR. I found similar issues documented in Dart SDK tooling, and applied a similar workaround.
    - Improve `debug: over_react_metrics` output to aid in performance optimizing
      - Star the top 3 slowest diagnostics
      - Show `getAllComponentUsages` overhead
    - `AnalyzerDebugHelper` - make log message and location computation lazy by default, so that we're not slowing down performance when debug flags aren't enabled
      - Update existing usages, add locations in missing_cascade_parens.dart
    
    

#### Release Notes
  

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
        - Manually verify that diagnostic behaves as expected in tools/analyzer_plugin/playground/web/missing_required_props.dart
        - Remove the `over_react_annotation_required_prop` line in the playground's analysis_options.yaml and verify that annotation-required props don't lint by default
        - Try typing in the playground and verify there are virtually no `InconsistentAnalysisExceptions` errors compared to the base branch
        - Verify behavior of debug flag comments looks good:
            - `// debug: over_react_metrics` 
                - Verify new behavior mentioned in Changes section looks good
                - Verify the performance of this diagnostic looks good
            - `// debug: over_react_required_props` - shows information about required props for each builder

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:


[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria
